TideApp/TideApp/Scenes/TideTimesView/TideTimesViewModel.swift

```diff
@@ -0,0 +1,95 @@
+//
+//  TideTimesViewModel.swift
+//  TideApp
+//
+//  Created by John Sanderson on 19/05/2021.
+//
+
+import Foundation
+import Combine
+import CoreLocation
+
+enum TideTimesState {
+  case idle
+  case loading
+  case error(Error)
+  case success(WeatherInfo)
+}
+
+protocol TideTimesViewModelType: ObservableObject {
+
+  var state: TideTimesState { get }
+
+  func getTideTimes()
+}
+
+// MARK: - Network ViewModel -
+
+/// A `TideTimesViewModelType` that fetches the `WeatherInfo` from the network using the provided location
+final class TideTimesNetworkViewModel: TideTimesViewModelType {
+
+  // MARK: - Properties -
+  // MARK: Internal
+
+  @Published var state: TideTimesState = .idle
+
+  // MARK: Private
+
+  private let location: CLLocation?
+  private let networkManager: NetworkManagerType
+  private var cancellable: AnyCancellable?
+
+  // MARK: - Initialiser -
+
+  init(location: CLLocation?, networkManager: NetworkManagerType = NetworkManager()) {
+    self.location = location
+    self.networkManager = networkManager
+  }
+
+  // MARK: - Functions -
+  // MARK: Internal
+
+  func getTideTimes() {
+    state = .loading
+    guard let location = location else {
+      state = .error(WeatherError.parsing(description: "Invalid location"))
+      return
+    }
+    let request = GetWeatherInformationRequest(location: location, date: Date(), networkManager: networkManager)
+    cancellable = request.perform()
+      .sink(receiveCompletion: { completion in
+        guard case let .failure(error) = completion else { return }
+        self.state = .error(error)
+      }, receiveValue: {
+        self.state = .success($0)
+      })
+  }
+}
+
+// MARK: - Local ViewModel -
+
+/// A `TideTimesViewModelType` that is initialised directly with a `WeatherInfo` object
+final class TideTimesLocalViewModel: TideTimesViewModelType {
```

Comment:
I like also this solution, but wouldn't be easier just have the possibility to instantiate the previous viewModel with the WeatherInfo or with the location + manager? This is really a question that I ask to myself too, its just curiosity again :D

Reply:
Yeah, either would work to be honest. Personally I think this way is a little cleaner as it avoids any logic of checking for optional weather info objects etc. **If** we were writing tests (😬) then I think this way would also be easier to test

Reply:
Yeah, make sense!!